src/Azure/Orleans.AzureCosmos/AzureCosmosGrainStorage.cs

```diff
@@ -0,0 +1,246 @@
+using Newtonsoft.Json;
+using Orleans.Configuration;
+using Orleans.Configuration.Overrides;
+using Orleans.Runtime;
+using Orleans.Serialization;
+using Orleans.Storage;
+using LogLevel = Microsoft.Extensions.Logging.LogLevel;
+
+namespace Orleans.AzureCosmos
+{
+    internal sealed class AzureCosmosGrainStorage : AzureCosmosStorage, IGrainStorage, ILifecycleParticipant<ISiloLifecycle>, ILifecycleObserver
+    {
+        private readonly AzureCosmosStorageOptions options;
+        private readonly string partitionPrefix;
+        private readonly IServiceProvider services;
+        private readonly string name;
+        private JsonSerializer serializer;
+
+        public static IGrainStorage Create(IServiceProvider sp, string name)
+            => ActivatorUtilities.CreateInstance<AzureCosmosGrainStorage>(sp, name, sp.GetProviderClusterOptions(name));
+
+        public AzureCosmosGrainStorage(
+            string name,
+            IOptionsMonitor<AzureCosmosStorageOptions> optionsSnapshot,
+            IOptions<ClusterOptions> clusterOptions,
+            IServiceProvider services,
+            ILoggerFactory loggerFactory)
+            : base(loggerFactory)
+        {
+            this.options = optionsSnapshot.Get(name);
+            this.partitionPrefix = clusterOptions.Value.ServiceId + "/";
+            this.name = name;
+            this.services = services;
+        }
+
+        public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            try
+            {
+                if (container == null) throw new InvalidOperationException("Uninitialized");
+
+                var pk = GetKeyString(grainReference);
+                if (logger.IsEnabled(LogLevel.Trace)) logger.Trace("Reading: GrainType={0} PK={1} GrainId={2} from Container={3}", grainType, pk, grainReference, options.ContainerName);
+
+                var startTime = DateTime.UtcNow;
+                // Task.Run is a workaround for https://github.com/Azure/azure-cosmos-dotnet-v2/issues/687
+                using var res = await Task.Run(() => container.ReadItemStreamAsync(grainType, new PartitionKey(pk)));
+                CheckAlertSlowAccess(startTime, "ReadItem");
+
+                if (res.StatusCode == HttpStatusCode.NotFound)
+                {
+                    if (logger.IsEnabled(LogLevel.Trace)) logger.Trace("NotFound reading: PartitionKey={0} RowKey={1} from Container={2}", pk, grainType, options.ContainerName);
+                    return;
+                }
+
+                var eTag = res.EnsureSuccessStatusCode().Headers.ETag;
+                if (logger.IsEnabled(LogLevel.Trace)) logger.Trace("Read: PartitionKey={0} RowKey={1} from Container={2} with ETag={3}", pk, grainType, options.ContainerName, eTag);
+
+                var record = (GrainStateRecord)serializer.Deserialize(new StreamReader(res.Content), typeof(GrainStateRecord<>).MakeGenericType(grainState.Type));
```

Comment:
How cheap is this MakeGenericType call, does it make sense to cache the results?